File: bindings/python/benchmarks/ffi_overhead.py
#!/usr/bin/env python3
"""
FFI Overhead Benchmarks for HEDL Python Bindings

Measures the performance overhead of FFI calls compared to native Rust operations.
Tests parse, convert, and validate operations across multiple document sizes.

Run with:
    python3 -m timeit -n 100 -r 5 -s "..." "..."
    or use: python3 benchmarks/ffi_overhead.py
"""

import timeit
import sys
import json
from typing import Dict, List, Tuple
import os

# Add parent directory to path for imports
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))

import hedl


# Test data generators
def generate_small_hedl() -> str:
    """Generate small HEDL document."""
    return """%VERSION: 1.0
%STRUCT: User: [id, name, email]
---
users: @User
  | 1, Alice, alice@example.com
  | 2, Bob, bob@example.com
"""


def generate_medium_hedl() -> str:
    """Generate medium HEDL document."""
    lines = ["%VERSION: 1.0", "%STRUCT: User: [id, name, email, dept]", "---", "users: @User"]
    for i in range(100):
        lines.append(f"  | {i}, User{i}, user{i}@example.com, dept{i%10}")
    return "\n".join(lines)


def generate_large_hedl() -> str:
    """Generate large HEDL document."""
    lines = ["%VERSION: 1.0", "%STRUCT: User: [id, name, email, dept, salary]", "---", "users: @User"]
    for i in range(1000):
        lines.append(f"  | {i}, User{i}, user{i}@example.com, dept{i%10}, {50000 + i*100}")
    return "\n".join(lines)


class BenchmarkResult:
    """Stores benchmark results for a single operation."""

    def __init__(self, name: str, operation: str, size: str):
        self.name = name
        self.operation = operation
        self.size = size
        self.times: List[float] = []
        self.overhead_percent = 0.0

    def add_time(self, time_ns: float):
        """Add a measurement in nanoseconds."""
        self.times.append(time_ns)

    def avg_time_ns(self) -> float:
        """Get average time in nanoseconds."""
        return sum(self.times) / len(self.times) if self.times else 0

    def min_time_ns(self) -> float:
        """Get minimum time in nanoseconds."""
        return min(self.times) if self.times else 0

    def max_time_ns(self) -> float:
        """Get maximum time in nanoseconds."""
        return max(self.times) if self.times else 0

    def std_dev_ns(self) -> float:
        """Get standard deviation in nanoseconds."""
        if len(self.times) < 2:
            return 0
        avg = self.avg_time_ns()
        variance = sum((t - avg) ** 2 for t in self.times) / (len(self.times) - 1)
        return variance ** 0.5

    def to_dict(self) -> Dict:
        """Convert to dictionary for JSON serialization."""
        return {
            "name": self.name,
            "operation": self.operation,
            "size": self.size,
            "avg_time_ns": round(self.avg_time_ns(), 2),
            "min_time_ns": round(self.min_time_ns(), 2),
            "max_time_ns": round(self.max_time_ns(), 2),
            "std_dev_ns": round(self.std_dev_ns(), 2),
            "overhead_percent": round(self.overhead_percent, 2),
            "samples": len(self.times),
        }


class BenchmarkSuite:
    """Runs FFI overhead benchmarks."""

    def __init__(self):
        self.results: List[BenchmarkResult] = []

    def benchmark_parse(self, hedl_content: str, size: str, iterations: int = 10):
        """Benchmark parse operation."""
        print(f"\nBenchmarking parse ({size})...")

        def parse_op():
            doc = hedl.parse(hedl_content)
            doc.close()

        times = []
        for _ in range(iterations):
            start = timeit.default_timer()
            parse_op()
            elapsed = (timeit.default_timer() - start) * 1e9  # Convert to ns
            times.append(elapsed)

        result = BenchmarkResult("parse", "Parse HEDL", size)
        for t in times:
            result.add_time(t)
        self.results.append(result)
        print(f"  Average: {result.avg_time_ns():.0f} ns")

    def benchmark_to_json(self, hedl_content: str, size: str, iterations: int = 10):
        """Benchmark to_json conversion."""
        print(f"\nBenchmarking to_json ({size})...")

        doc = hedl.parse(hedl_content)

        def to_json_op():
            return doc.to_json()

        times = []
        try:
            for _ in range(iterations):
                start = timeit.default_timer()
                to_json_op()
                elapsed = (timeit.default_timer() - start) * 1e9  # Convert to ns
                times.append(elapsed)

            result = BenchmarkResult("to_json", "Convert to JSON", size)
            for t in times:
                result.add_time(t)
            self.results.append(result)
            print(f"  Average: {result.avg_time_ns():.0f} ns")
        finally:
            doc.close()

    def benchmark_to_yaml(self, hedl_content: str, size: str, iterations: int = 10):
        """Benchmark to_yaml conversion."""
        print(f"\nBenchmarking to_yaml ({size})...")

        doc = hedl.parse(hedl_content)

        def to_yaml_op():
            return doc.to_yaml()

        times = []
        try:
            for _ in range(iterations):
                start = timeit.default_timer()
                to_yaml_op()
                elapsed = (timeit.default_timer() - start) * 1e9  # Convert to ns
                times.append(elapsed)

            result = BenchmarkResult("to_yaml", "Convert to YAML", size)
            for t in times:
                result.add_time(t)
            self.results.append(result)
            print(f"  Average: {result.avg_time_ns():.0f} ns")
        finally:
            doc.close()

    def benchmark_validate(self, hedl_content: str, size: str, iterations: int = 10):
        """Benchmark validate operation."""
        print(f"\nBenchmarking validate ({size})...")

        def validate_op():
            return hedl.validate(hedl_content)

        times = []
        for _ in range(iterations):
            start = timeit.default_timer()
            validate_op()
            elapsed = (timeit.default_timer() - start) * 1e9  # Convert to ns
            times.append(elapsed)

        result = BenchmarkResult("validate", "Validate HEDL", size)
        for t in times:
            result.add_time(t)
        self.results.append(result)
        print(f"  Average: {result.avg_time_ns():.0f} ns")

    def run_all(self):
        """Run all benchmarks."""
        print("=" * 70)
        print("HEDL Python FFI Overhead Benchmarks")
        print("=" * 70)

        test_cases = [
            ("small", generate_small_hedl(), 20),
            ("medium", generate_medium_hedl(), 10),
            ("large", generate_large_hedl(), 5),
        ]

        for size_name, content, iterations in test_cases:
            print(f"\n{'='*70}")
            print(f"Testing {size_name} documents ({len(content)} bytes)")
            print(f"{'='*70}")

            self.benchmark_parse(content, size_name, iterations)
            self.benchmark_validate(content, size_name, iterations)
            self.benchmark_to_json(content, size_name, iterations)
            self.benchmark_to_yaml(content, size_name, iterations)

    def print_summary(self):
        """Print summary of all benchmarks."""
        print("\n" + "=" * 70)
        print("BENCHMARK SUMMARY")
        print("=" * 70)

        # Group by operation
        operations = {}
        for result in self.results:
            key = result.operation
            if key not in operations:
                operations[key] = []
            operations[key].append(result)

        for op_name, op_results in sorted(operations.items()):
            print(f"\n{op_name}:")
            print(f"  {'Size':<10} {'Avg (ns)':<15} {'Min (ns)':<15} {'Max (ns)':<15} {'StdDev':<15}")
            print("  " + "-" * 70)
            for result in sorted(op_results, key=lambda r: int(r.size == "small") or int(r.size == "medium") or 2):
                print(
                    f"  {result.size:<10} {result.avg_time_ns():<15.0f} "
                    f"{result.min_time_ns():<15.0f} {result.max_time_ns():<15.0f} "
                    f"{result.std_dev_ns():<15.0f}"
                )

    def export_json(self, filename: str = "ffi_overhead_results.json"):
        """Export results to JSON file."""
        data = {
            "benchmark": "HEDL Python FFI Overhead",
            "timestamp": __import__("datetime").datetime.now().isoformat(),
            "results": [r.to_dict() for r in self.results],
        }

        with open(filename, "w") as f:
            json.dump(data, f, indent=2)
        print(f"\nResults exported to {filename}")


def main():
    """Run benchmarks."""
    suite = BenchmarkSuite()
    suite.run_all()
    suite.print_summary()
    suite.export_json()


if __name__ == "__main__":
    main()

File: bindings/php/benchmarks/ffi_overhead.php
<?php
/**
 * FFI Overhead Benchmarks for HEDL PHP Bindings
 *
 * Measures the performance overhead of FFI calls compared to native Rust operations.
 * Tests parse, convert, and validate operations across multiple document sizes.
 *
 * Requirements:
 * - PHP 8.0+ with FFI extension enabled
 * - libhedl_ffi shared library
 *
 * Run with:
 *   php benchmarks/ffi_overhead.php
 *   or: time php benchmarks/ffi_overhead.php
 */

namespace Dweve\Hedl\Benchmarks;

require_once __DIR__ . '/../Hedl.php';

use Dweve\Hedl\Hedl;

// Test data generators
function generateSmallHedl(): string
{
    return <<<'HEDL'
%VERSION: 1.0
%STRUCT: User: [id, name, email]
---
users: @User
  | 1, Alice, alice@example.com
  | 2, Bob, bob@example.com
HEDL;
}

function generateMediumHedl(): string
{
    $lines = [
        '%VERSION: 1.0',
        '%STRUCT: User: [id, name, email, dept]',
        '---',
        'users: @User'
    ];
    for ($i = 0; $i < 100; $i++) {
        $lines[] = "  | $i, User$i, user$i@example.com, dept" . ($i % 10);
    }
    return implode("\n", $lines);
}

function generateLargeHedl(): string
{
    $lines = [
        '%VERSION: 1.0',
        '%STRUCT: User: [id, name, email, dept, salary]',
        '---',
        'users: @User'
    ];
    for ($i = 0; $i < 1000; $i++) {
        $salary = 50000 + $i * 100;
        $lines[] = "  | $i, User$i, user$i@example.com, dept" . ($i % 10) . ", $salary";
    }
    return implode("\n", $lines);
}

/**
 * Stores benchmark results for a single operation
 */
class BenchmarkResult
{
    private string $name;
    private string $operation;
    private string $size;
    private array $times = [];
    private float $overheadPercent = 0.0;

    public function __construct(string $name, string $operation, string $size)
    {
        $this->name = $name;
        $this->operation = $operation;
        $this->size = $size;
    }

    public function addTime(float $timeNs): void
    {
        $this->times[] = $timeNs;
    }

    public function avgTimeNs(): float
    {
        return empty($this->times) ? 0.0 : array_sum($this->times) / count($this->times);
    }

    public function minTimeNs(): float
    {
        return empty($this->times) ? 0.0 : min($this->times);
    }

    public function maxTimeNs(): float
    {
        return empty($this->times) ? 0.0 : max($this->times);
    }

    public function stdDevNs(): float
    {
        if (count($this->times) < 2) {
            return 0.0;
        }
        $avg = $this->avgTimeNs();
        $variance = array_sum(array_map(fn($t) => pow($t - $avg, 2), $this->times)) / (count($this->times) - 1);
        return sqrt($variance);
    }

    public function toArray(): array
    {
        return [
            'name' => $this->name,
            'operation' => $this->operation,
            'size' => $this->size,
            'avg_time_ns' => round($this->avgTimeNs(), 2),
            'min_time_ns' => round($this->minTimeNs(), 2),
            'max_time_ns' => round($this->maxTimeNs(), 2),
            'std_dev_ns' => round($this->stdDevNs(), 2),
            'overhead_percent' => round($this->overheadPercent, 2),
            'samples' => count($this->times),
        ];
    }

    public function getName(): string
    {
        return $this->name;
    }

    public function getOperation(): string
    {
        return $this->operation;
    }

    public function getSize(): string
    {
        return $this->size;
    }
}

/**
 * Runs FFI overhead benchmarks
 */
class BenchmarkSuite
{
    private array $results = [];

    public function benchmarkParse(string $content, string $size, int $iterations = 10): void
    {
        printf("\nBenchmarking parse (%s)...\n", $size);

        $times = [];
        for ($i = 0; $i < $iterations; $i++) {
            $start = hrtime(true);
            $doc = Hedl::parse($content);
            $doc->close();
            $elapsed = hrtime(true) - $start;
            $times[] = $elapsed;
        }

        $result = new BenchmarkResult('parse', 'Parse HEDL', $size);
        foreach ($times as $t) {
            $result->addTime($t);
        }
        $this->results[] = $result;
        printf("  Average: %.0f ns\n", $result->avgTimeNs());
    }

    public function benchmarkToJson(string $content, string $size, int $iterations = 10): void
    {
        printf("\nBenchmarking to_json (%s)...\n", $size);

        $doc = Hedl::parse($content);

        $times = [];
        for ($i = 0; $i < $iterations; $i++) {
            $start = hrtime(true);
            $doc->toJson();
            $elapsed = hrtime(true) - $start;
            $times[] = $elapsed;
        }

        $doc->close();

        $result = new BenchmarkResult('to_json', 'Convert to JSON', $size);
        foreach ($times as $t) {
            $result->addTime($t);
        }
        $this->results[] = $result;
        printf("  Average: %.0f ns\n", $result->avgTimeNs());
    }

    public function benchmarkToYaml(string $content, string $size, int $iterations = 10): void
    {
        printf("\nBenchmarking to_yaml (%s)...\n", $size);

        $doc = Hedl::parse($content);

        $times = [];
        for ($i = 0; $i < $iterations; $i++) {
            $start = hrtime(true);
            $doc->toYaml();
            $elapsed = hrtime(true) - $start;
            $times[] = $elapsed;
        }

        $doc->close();

        $result = new BenchmarkResult('to_yaml', 'Convert to YAML', $size);
        foreach ($times as $t) {
            $result->addTime($t);
        }
        $this->results[] = $result;
        printf("  Average: %.0f ns\n", $result->avgTimeNs());
    }

    public function benchmarkValidate(string $content, string $size, int $iterations = 10): void
    {
        printf("\nBenchmarking validate (%s)...\n", $size);

        $times = [];
        for ($i = 0; $i < $iterations; $i++) {
            $start = hrtime(true);
            Hedl::validate($content);
            $elapsed = hrtime(true) - $start;
            $times[] = $elapsed;
        }

        $result = new BenchmarkResult('validate', 'Validate HEDL', $size);
        foreach ($times as $t) {
            $result->addTime($t);
        }
        $this->results[] = $result;
        printf("  Average: %.0f ns\n", $result->avgTimeNs());
    }

    public function runAll(): void
    {
        echo str_repeat('=', 70) . "\n";
        echo "HEDL PHP FFI Overhead Benchmarks\n";
        echo str_repeat('=', 70) . "\n";

        $testCases = [
            ['small', generateSmallHedl(), 20],
            ['medium', generateMediumHedl(), 10],
            ['large', generateLargeHedl(), 5],
        ];

        foreach ($testCases as [$sizeName, $content, $iterations]) {
            printf("\n%s\n", str_repeat('=', 70));
            printf("Testing %s documents (%d bytes)\n", $sizeName, strlen($content));
            printf("%s\n", str_repeat('=', 70));

            $this->benchmarkParse($content, $sizeName, $iterations);
            $this->benchmarkValidate($content, $sizeName, $iterations);
            $this->benchmarkToJson($content, $sizeName, $iterations);
            $this->benchmarkToYaml($content, $sizeName, $iterations);
        }
    }

    public function printSummary(): void
    {
        echo "\n" . str_repeat('=', 70) . "\n";
        echo "BENCHMARK SUMMARY\n";
        echo str_repeat('=', 70) . "\n";

        $operations = [];
        foreach ($this->results as $result) {
            $op = $result->getOperation();
            if (!isset($operations[$op])) {
                $operations[$op] = [];
            }
            $operations[$op][] = $result;
        }

        $sizeOrder = ['small' => 0, 'medium' => 1, 'large' => 2];

        ksort($operations);
        foreach ($operations as $opName => $opResults) {
            printf("\n%s:\n", $opName);
            printf("  %-10s %-15s %-15s %-15s %-15s\n", 'Size', 'Avg (ns)', 'Min (ns)', 'Max (ns)', 'StdDev');
            echo "  " . str_repeat('-', 70) . "\n";

            usort($opResults, fn($a, $b) => ($sizeOrder[$a->getSize()] ?? 999) <=> ($sizeOrder[$b->getSize()] ?? 999));

            foreach ($opResults as $result) {
                printf(
                    "  %-10s %-15.0f %-15.0f %-15.0f %-15.0f\n",
                    $result->getSize(),
                    $result->avgTimeNs(),
                    $result->minTimeNs(),
                    $result->maxTimeNs(),
                    $result->stdDevNs()
                );
            }
        }
    }

    public function exportJson(string $filename = 'ffi_overhead_results.json'): void
    {
        $data = [
            'benchmark' => 'HEDL PHP FFI Overhead',
            'timestamp' => (new \DateTime())->format(\DateTime::RFC3339),
            'results' => array_map(fn($r) => $r->toArray(), $this->results),
        ];

        file_put_contents($filename, json_encode($data, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES));
        printf("\nResults exported to %s\n", $filename);
    }
}

function main(): void
{
    $suite = new BenchmarkSuite();
    $suite->runAll();
    $suite->printSummary();
    $suite->exportJson();
}

main();
